### PR TITLE
tests: Use watch events to patch history to speed up linearization

### PR DESCRIPTION
**Pull request by serathius**
cc @ahrtr @ptabor

### CODE REVIEW

File: tests/linearizability/linearizability_test.go [hunk 1]
```diff
@@ -133,6 +136,62 @@ func testLinearizability(ctx context.Context, t *testing.T, clus *e2e.EtcdProces
 	return operations, events
 }
 
+func patchOperationHistory(operations []porcupine.Operation, watchOps []watchEvent) []porcupine.Operation {
+	newOperations := make([]porcupine.Operation, 0, len(operations))
+	persisted := map[EtcdOperation]watchEvent{}
```
Comment (ptabor) [inline tests/linearizability/linearizability_test.go:141]: operationId2watch ?

File: tests/linearizability/linearizability_test.go [hunk 2]
```diff
@@ -133,6 +136,62 @@ func testLinearizability(ctx context.Context, t *testing.T, clus *e2e.EtcdProces
...
+		hasPut := false
+		wasPersisted := false
+		hasDelete := false
+		for _, op1 := range request.Ops {
+			switch op1.Type {
+			case Delete:
+				hasDelete = true
+			case Put:
+				hasPut = true
+				op2, ok := persisted[op1]
```
Comment (ptabor) [inline tests/linearizability/linearizability_test.go:167]: watchedOp, := ...

File: tests/linearizability/linearizability_test.go [hunk 3]
```diff
@@ -133,6 +136,62 @@ func testLinearizability(ctx context.Context, t *testing.T, clus *e2e.EtcdProces
...
+			newOperations = append(newOperations, op)
+			continue
+		}
+		hasPut := false
+		wasPersisted := false
+		hasDelete := false
+		for _, op1 := range request.Ops {
+			switch op1.Type {
+			case Delete:
+				hasDelete = true
```
Comment (ptabor) [inline tests/linearizability/linearizability_test.go:164]: I would comment here why we cannot do the same for deletes. (I see the comment below, about deletes being not unique... but here might be more natural place)

File: tests/linearizability/linearizability_test.go [hunk 4]
```diff
@@ -212,20 +271,22 @@ type trafficConfig struct {
 	traffic     Traffic
 }
 
-func validateEventsMatch(t *testing.T, ops [][]watchEvent) {
-	// Move longest history to ops[0]
+func pickLongestHistory(ops [][]watchEvent) (longest []watchEvent, rest [][]watchEvent) {
```
Comment (ptabor) [inline tests/linearizability/linearizability_test.go:274]: Just having 'sortDescByLenght' would have simpler contract and shorter implementation. 
Cost of sorting of O(3) pointers to lists is negligible. 

Now we have function that sounds like getter... but is mutating 'ops'. 

```
sort.Slice(ops, func(i, j []watchEvent) bool {
  return len(i)>len();
})
```

**Review by ptabor — APPROVED** [review]
Very good idea ! Just a few nits.

File: tests/linearizability/linearizability_test.go (hunk 1 above)
Comment (serathius) [reply tests/linearizability/linearizability_test.go:141]: Not very convinced. This function takes operations and patches them based on watch events.

Comment (serathius) [reply tests/linearizability/linearizability_test.go:141]: Renamed to `patchOperationBasedOnWatchEvents`

File: tests/linearizability/linearizability_test.go (hunk 2 above)
Comment (serathius) [reply tests/linearizability/linearizability_test.go:167]: Thanks for pointing this out. Forgot about this :P

File: tests/linearizability/linearizability_test.go (hunk 3 above)
Comment (serathius) [reply tests/linearizability/linearizability_test.go:164]: Done

**Comment by serathius** [comment]
Do not merge, still working on some weird interactions with leases that were just added.

**Pull request merged**